File: 0516_longestPalindromicSubsequence/main.cpp
#include <string>
#include <vector>
#include <algorithm>
// #include <bits/stdc++.h>

class Solution {
public:
    // Bottom-up approach
    // TC: O(N*N), SC: O(N*N)
    int longestPalindromeSubseq(std::string& s) {
        if(s.empty())
            return 0;
        
        const int N = s.size();
        std::vector<std::vector<int>> dp(N, std::vector<int>(N, 0));
        
        for(int i = 0; i < N; i++)
            dp[i][i] = 1;
        
        for(int l = 1; l < N; l++) {
            for(int i = 0; i < N - l; i++) {
                int j = i + l;
                
                // if there are only two chars
                if((j - i + 1) == 2) {
                    // individually they are palindromic so max length 1
                    dp[i][j] = 1 + (s[i] == s[j]);
                }
                else {
                    // for s[i:j]:
                    // longest palindrome length in s[i+1 : j-1] and +1 if s[i] == s[j]
                    // if s[i] != s[j], longest length palindrome = Max of longest palin in s[i:j-1] and s[i+1:j]
                    if(s[i] == s[j])
                        dp[i][j] = dp[i+1][j-1] + 2;
                    else
                        dp[i][j] = std::max(dp[i+1][j], dp[i][j-1]);
                }
            }
        }
        return dp[0][N-1];
    }
    

};

class Solution1 {
public:
    // Using memoization
    // TC: O(N*N), SC: O(N*N)
    int longestPalindSubseqRec(std::vector<std::vector<int> >& dp, std::string& s, int i, int j) {
        if(i > j)
            return 0;

        if(i == j)
            return dp[i][j] = 1;
        
        if(dp[i][j] == -1) {
            if(s[i] == s[j])
                dp[i][j] = 2 + longestPalindSubseqRec(dp, s, i+1, j-1);
            else
                dp[i][j] = std::max(longestPalindSubseqRec(dp, s, i+1, j),
                               longestPalindSubseqRec(dp, s, i, j-1));
        }
        return dp[i][j];
    }
    
    int longestPalindromeSubseq(std::string& s) {
        if(s.empty())
            return 0;

        const int N = s.size();
        // dp(i, j): length of longest palindromic substring in s[i:j]
        std::vector<std::vector<int> > dp(N, std::vector<int>(N, -1)); 
        
        return longestPalindSubseqRec(dp, s, 0, N-1);
    }
};

File: 0516_longestPalindromicSubsequence/main.py
class Solution:
    def longestPalindromeSubseq(self, s: str) -> int:
        reverse_s = s[::-1]
        
        dp=[[0 for _ in range(len(s)+1)] for __ in range(len(s)+1)]
        
        for i in range(1, len(s) + 1): 
            for j in range(1, len(s) + 1):
                if s[i-1] != reverse_s[j-1]:
                    dp[i][j]=max(dp[i][j-1], dp[i-1][j])
                else:
                    dp[i][j] = 1 + dp[i-1][j-1]
        return dp[-1][-1]

class Solution1:
    def longestPalindromeSubseq(self, s: str) -> int:
        n = len(s)
        dp = [[0] * n for _ in range(n)]
        
        for i in range(n):
            dp[i][i] = 1
            
        # Iterate over substrings of length 2 to n
        for l in range(2, n+1):
            for i in range(n-l+1):
                j = i + l - 1
                if s[i] == s[j]:
                    dp[i][j] = dp[i+1][j-1] + 2
                else:
                    dp[i][j] = max(dp[i+1][j], dp[i][j-1])
        
        return dp[0][n-1]
    
